cmd/http-tracer.go

```diff
@@ -206,7 +207,7 @@ func Trace(f http.HandlerFunc, logBody bool, w http.ResponseWriter, r *http.Requ
 
 	// Setup a http response body recorder
 	respBodyRecorder := &recordResponseWriter{ResponseWriter: w, logBody: logBody}
-	f(respBodyRecorder, r)
+	f(logger.NewResponseWriter(respBodyRecorder), r)
```

Comment:
The other place this is being done is:
```
func (s customHeaderHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
        // Set custom headers such as x-amz-request-id for each request.
        w.Header().Set(xhttp.AmzRequestID, mustGetRequestID(UTCNow()))
        s.handler.ServeHTTP(logger.NewResponseWriter(w), r)
}
```

do you know why `logger.NewResponseWriter(w)` was added in `customHeaderHandler` middleware because they seem to be unrelated.

Also adding `logger.NewResponseWriter(respBodyRecorder)` in `Trace()` is not correct since they are not related.

We can consolidate it to be done in one single place like:
```
diff --git a/cmd/api-router.go b/cmd/api-router.go
index 18ead4f3..f7676ab5 100644
--- a/cmd/api-router.go
+++ b/cmd/api-router.go
@@ -59,7 +59,7 @@ func registerAPIRouter(router *mux.Router, encryptionEnabled, allowSSEKMS bool)
        for _, bucket := range routers {
                // Object operations
                // HeadObject
-               bucket.Methods(http.MethodHead).Path("/{object:.+}").HandlerFunc(httpTraceAll(api.HeadObjectHandler))
+               bucket.Methods(http.MethodHead).Path("/{object:.+}").HandlerFunc(httpTraceAll(auditLog(api.HeadObjectHandler, "HeadObject")))
                // CopyObjectPart
                bucket.Methods(http.MethodPut).Path("/{object:.+}").HeadersRegexp(xhttp.AmzCopySource, ".*?(\\/|%2F).*?").HandlerFunc(httpTraceAll(api.CopyObjectPartHandler)).Queries("partNumber", "{partNumber:[0-9]+}", "uploadId", "{uploadId:.*}")
                // PutObjectPart
diff --git a/cmd/handler-utils.go b/cmd/handler-utils.go
index 981c739c..8ca70975 100644
--- a/cmd/handler-utils.go
+++ b/cmd/handler-utils.go
@@ -324,6 +324,19 @@ func extractPostPolicyFormValues(ctx context.Context, form *multipart.Form) (fil
        return filePart, fileName, fileSize, formValues, nil
 }
 
+func auditLog(f http.HandlerFunc, method string) http.HandlerFunc {
+       return func(w http.ResponseWriter, r *http.Request) {
+               f.ServeHTTP(logger.NewResponseWriter(w), r)
+               logger.AuditLog(w, r, method, mustGetClaimsFromToken(r))
+       }
+}
```

Reply:
The refactor you are asking is perhaps fine, but it cannot be done right now. I would like to take current fix and refactor this later. The minimal change that is needed right now is the PR that its fixing.

Audit logging is meant for all incoming calls - that is why it's not Router specific, but in-fact a more generic handler implementation.